### DISCUSSION TITLE
NFluxPerCell does not work

### DISCUSSION THREAD

**Discussion by deSUBHAM** [BLAST-WarpX/warpx#5344] [General]
Hi, 
I am making a setup for magnetic reconnection in pair plasma. The current sheet constitutes with particle having drift velocity. The background particles do not have drift velocity. I want to inject the background velocity. For that I used NFluxPerCell, as suggested here. But when I am doing that, the corresponding species is not detected by the code. The initialisation of background electrons are given below. 
```
electrons_b.species_type                           = "electron"
electrons_b.injection_style                        = "NFluxPerCell"
electrons_b.surface_flux_pos                       = -Lz
electrons_b.flux_normal_axis                       = z
electrons_b.flux_direction                         = -1
electrons_b.num_particles_per_cell                 = 128
electrons_b.profile                                = "constant"
electrons_b.density                                = nb
electrons_b.momentum_distribution_type             = "gaussianflux"
#electrons_b.theta_distribution_type                = "constant"
#electrons_b.theta                                  = temp_b
electrons_b.ux_m                                   = clight * sqrt((2*temp_b)/pi)
electrons_b.ux_th                                  = clight * sqrt(temp_b)
electrons_b.uy_m                                   = clight * sqrt((2*temp_b)/pi)
electrons_b.uy_th                                  = clight * sqrt(temp_b)
electrons_b.uz_m                                   = clight * sqrt((2*temp_b)/pi)
electrons_b.uz_th                                  = clight * sqrt(temp_b)
```
I checked with Maxwell-Boltzmann distribution distribution, it did not work. Someone please help me out here. Thank you.

**Comment by dpgrote** [comment]
Is the `surface_flux_pos = -Lz` the lower bound of the domain? I notice that `flux_direction = -1` which means that the particles will be injected with a negative velocity and never enter the domain. Also, are those typos in the expressions `2temp_b`? Should those be `2*temp_b`?

**Reply by deSUBHAM** [reply]
Thank you so much. I had a bit confusion about 'flux_direction'. Now it is clear. Also I do not find any typo of `2temp_b` in the part of the code. 
However, I have a couple of doubts though. 
1. For the injection style "NFluxPerCell", does one have to have the momentum distribution style  "gaussianflux"? I require the Maxwell_boltzmann distribution of momentum. 
2. In my reconnection setup, it appears that the background particles are not participating in the reconnection process if they are injected as "NFluxPerCell". Otherwise, there seems to be no problem if they are injected as "RandomPerCell". Is there a special way to initialise them or am I doing something wrong?

**Reply by dpgrote** [reply]
1 The `gaussianflux` is what you want. If this detail is unclear in the documentation, can you make a recommendation about how to change the documentation to make it more clear?
2 Can you check the particle properties, the charge, mass, and weight, to make sure that they are initialized correctly?

**Comment by deSUBHAM** [comment]
Hi, 
Please find the attached input file (input_2d_SM.txt). Here, I am trying to run for a 2d asymmetric magnetic reconnection case. The boundary conditions along x-direction is "absorbing_silver_mueller" for fields and "Abosorbing" for particles. The z-direction is periodic, as you can see it in the inout file. I have several queries based on that. These are listed below.

1. Since the particles are leaving the domain, I want to inject particles apart from the initial particles, which have drift velocity to provide the current. Specifically, I want to inject particles from the two ends of the x-boundaries. For that, I used "NFluxPerCell" with momentum_distribution_type = "gaussianflux" and "maxwell_boltzmann". However, in both the cases, the number density of the injected particles (~0.13) is a few orders of magnitude smaller than the expected ones (~10^7). One of the snapshot (NFluxPerCell.pdf) is attached here. I would like to use "NFluxPerCell" instead of using the moving window, the former has more parameters to control, as I realise. It would be nice if someone clarify the issue and help me resolve it. 
2. I also have query about whether I can use "maxwell_boltzmann" distribution in "NFluxPerCell". 
3. I would like to have another clarification. I tried with PEC and reflecting boundary conditions for fields and particles respectively. However, at some point, the total energy (particle + EM) increases due to the generation of some electric fields near the boundaries. I think this issue is mentioned in the documentation. Nevertheless, if I use absorbing boundary conditions (Silver-Muller and/or PML) on fields, the field values quickly tend to zero at the boundaries. Please refer to the attached figure (B_n.pdf). It would be nice to know which boundary condition to use so that the total energy remains constant and field remains more or less similar until they reconnect. If we have to inject field from the boundaries, that is also okay. But, please let me know how to keep the magnetic field sustained. 

Thank you so much.

[B_n.pdf](https://github.com/user-attachments/files/21755905/B_n.pdf)
[input_2d_SM.txt](https://github.com/user-attachments/files/21754426/input_2d_SM.txt)
[NFluxPerCell.pdf](https://github.com/user-attachments/files/21754726/NFluxPerCell.pdf)

**Comment by aeriforme** [comment]
About 1, I don't see `<species_name>.flux_profile` anywhere in the input file, instead I see the `<species_name>.profile`. 
I don't think the two are (or should be) compatible. Did you get any warning?
In any case, to inject a flux of particles, one needs to define the value of the flux as the number of particles per unit surface and unit time (with `flux_profile`), not the number density (`profile`). Does this make sense?
 
About 2, I think you can, yes. The `gaussianflux` is an additional option for momentum initialization in case of injection via `NFluxPerCell`.
